src/main/java/net/minestom/server/listener/PlayerDiggingListener.java

```diff
@@ -102,10 +102,23 @@ private static DiggingResult cancelDigging(Player player, Instance instance, Poi
     private static DiggingResult finishDigging(Player player, Instance instance, Point blockPosition, BlockFace blockFace) {
         final Block block = instance.getBlock(blockPosition);
 
+        System.out.println(" Packet Finish Digging");
+
         if (shouldPreventBreaking(player, block)) {
             return new DiggingResult(block, false);
         }
 
+        final int breakTicks = BlockBreakCalculation.breakTicks(block, player);
+        // Realistically shouldn't happen, but a hacked client can send any packet, also illegal ones
+        // If the block is unbreakable, prevent a hacked client from breaking it!
+        if (breakTicks == BlockBreakCalculation.UNBREAKABLE) {
+            PlayerCancelDiggingEvent playerCancelDiggingEvent = new PlayerCancelDiggingEvent(player, block, new BlockVec(blockPosition));
+            EventDispatcher.call(playerCancelDiggingEvent);
+            return new DiggingResult(block, false);
+        }
+        // TODO maybe add a check if the player has spent enough time mining the block.
```

Comment:
I don't believe its Minestoms job to check this.

Reply:
Actually, I believe it is. Basic illegal behaviour should be checked, that includes a sent FINISHED_DIGGING packet on an unbreakable block. Otherwise players can go around breaking any block (bedrock, barrier) they like if the break isn't cancelled by the server owner.
Similar sanity checks are already there, IMO not harmful adding another:
https://github.com/Minestom/Minestom/blob/480a7dcaa2b26a074c97583f8177d2f7a3677876/src/main/java/net/minestom/server/listener/PlayerDiggingListener.java#L115

Reply:
vanilla has a similar time check, which is why blocks pop up again after broken if the server is lagging or you have a bad internet connection. I haven't added the check because it is not trivial to do it right, I just left the comment because it is something that could be considered in the future.